flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/ColumnarArray.java

```diff
@@ -0,0 +1,332 @@
+
+	private final DataType elementType;
+	private final ColumnVector data;
+	private final int offset;
+	private final int length;
+
+	public ColumnarArray(DataType elementType, ColumnVector data, int offset, int length) {
+		this.elementType = elementType;
+		this.data = data;
+		this.offset = offset;
+		this.length = length;
+	}
+
+	@Override
+	public int numElements() {
+		return length;
+	}
+
+	@Override
+	public boolean isNullAt(int pos) {
+		return data.isNullAt(offset + pos);
+	}
+
+	@Override
+	public void setNullAt(int pos) {
+		throw new UnsupportedOperationException("Not support the operation!");
+	}
+
+	@Override
+	public boolean getBoolean(int ordinal) {
+		return ((BooleanColumnVector) data).getBoolean(offset + ordinal);
+	}
+
+	@Override
+	public byte getByte(int ordinal) {
+		return ((ByteColumnVector) data).getByte(offset + ordinal);
+	}
+
+	@Override
+	public short getShort(int ordinal) {
+		return ((ShortColumnVector) data).getShort(offset + ordinal);
+	}
+
+	@Override
+	public int getInt(int ordinal) {
+		return ((IntColumnVector) data).getInt(offset + ordinal);
+	}
+
+	@Override
+	public long getLong(int ordinal) {
+		return ((LongColumnVector) data).getLong(offset + ordinal);
+	}
+
+	@Override
+	public float getFloat(int ordinal) {
+		return ((FloatColumnVector) data).getFloat(offset + ordinal);
+	}
+
+	@Override
+	public double getDouble(int ordinal) {
+		return ((DoubleColumnVector) data).getDouble(offset + ordinal);
+	}
+
+	@Override
+	public BinaryString getString(int ordinal) {
+		BytesColumnVector.Bytes byteArray = getByteArray(ordinal);
+		return BinaryString.fromBytes(byteArray.data, byteArray.offset, byteArray.len);
+	}
+
+	@Override
+	public Decimal getDecimal(int ordinal, int precision, int scale) {
+		return ((DecimalColumnVector) data).getDecimal(offset + ordinal, precision, scale);
+	}
+
+	@Override
+	public SqlTimestamp getTimestamp(int ordinal, int precision) {
+		return ((TimestampColumnVector) data).getTimestamp(offset + ordinal, precision);
+	}
+
+	@Override
+	public <T> BinaryGeneric<T> getGeneric(int ordinal) {
+		throw new UnsupportedOperationException("GenericType is not supported.");
+	}
+
+	@Override
+	public byte[] getBinary(int ordinal) {
+		BytesColumnVector.Bytes byteArray = getByteArray(ordinal);
+		if (byteArray.len == byteArray.data.length) {
+			return byteArray.data;
+		} else {
+			byte[] ret = new byte[byteArray.len];
+			System.arraycopy(byteArray.data, byteArray.offset, ret, 0, byteArray.len);
+			return ret;
+		}
+	}
+
+	@Override
+	public BaseArray getArray(int ordinal) {
+		return ((ArrayColumnVector) data).getArray(offset + ordinal);
+	}
+
+	@Override
+	public BaseMap getMap(int ordinal) {
+		throw new UnsupportedOperationException("Map is not supported.");
+	}
+
+	@Override
+	public BaseRow getRow(int ordinal, int numFields) {
+		throw new UnsupportedOperationException("Row is not supported.");
+	}
+
+	@Override
+	public void setBoolean(int ordinal, boolean value) {
```

Comment:
These annoyed "set" will be removed after FLIP-95.

Reply:
Thanks for letting me know. As this is still needed for now and so just keep it in this PR and we can remove it during FLIP-95.